File: 백준/Silver/9711. 피보나치/피보나치.py
import sys
input = sys.stdin.readline


tc = int(input().strip())
pibo = [0 for _ in range(10001)]
for k in range(tc):
    P, Q = map(int, input().split())
    pibo[0] = pibo[1] = 1
    for i in range(2, P):
        if not pibo[i]:
            pibo[i] = pibo[i-1] + pibo[i-2]
    
    print(f'Case #{k+1}: {pibo[P-1]%Q}')

File: 백준/Silver/9711. 피보나치/피보나치.ts
const fs = require("fs");
const stdin = (
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString()
    : `10
    5 10
    6 25
    10 21
    32 43
    100 100
    50 50
    25 25
    45 67
    109 32
    128 128
`
)
  .trim()
  .split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const testCase : number = parseInt(input())
const pibo : bigint[] = Array(10001).fill(BigInt(0))
for (let tc = 0; tc < testCase; tc++) {
  const [P, Q] : number[] = input().trim().split(" ").map(Number)
  const q = BigInt(Q)
  pibo[0] = BigInt(1)
  pibo[1] = BigInt(1)
  for (let i = 2; i < P; i++) {
    if (pibo[i] === BigInt(0)) {
      pibo[i] = pibo[i-2] + pibo[i-1]
    }
  }
  console.log(`Case #${tc+1}: ${Number((pibo[P-1]%q))}`)
}


